In [1]:
# 🔥 STEP 1 – Project Setup (Very Important Foundation)

In [2]:
# 🔥 STEP 2 – Importing Libraries & Loading Data

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Display all columns
pd.set_option('display.max_columns', None)

In [4]:
df = pd.read_excel("../data/Online Retail.xlsx")

df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [5]:
df.tail()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France
541908,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,2011-12-09 12:50:00,4.95,12680.0,France


In [6]:
df.shape

(541909, 8)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [8]:
# 🔥 STEP 3 – Check Missing Values Properly

In [9]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [10]:
# 🔥 STEP 4 – Data Cleaning Strategy (Explain Before Code)

In [11]:
# 🔥 STEP 5 – Start Cleaning (Controlled & Measured)

In [12]:
original_rows = df.shape[0]
original_rows

541909

In [13]:
# 🔥 STEP 6 – Remove Missing CustomerID

In [14]:
# Droping the Customerid rows which has been empty because for accuraccy (almost 25% have been missing)
df = df.dropna(subset=['CustomerID'])

In [15]:
after_customer_id_drop = df.shape[0]
after_customer_id_drop

406829

In [16]:
removed_customer_rows = original_rows - after_customer_id_drop
removed_customer_rows

135080

In [17]:
# Removed ~25% due to missing customers
(removed_customer_rows / original_rows) * 100

24.926694334288598

In [18]:
# 🔥 STEP 7 – Remove Cancellations

In [19]:
# Used to check Invoice-no that starts with "C" means cancelled orders
df['InvoiceNo'].astype(str).str.startswith('C').sum()

8905

In [20]:
# 🔥 STEP 8 – Remove Cancellations

In [21]:
df = df[~df['InvoiceNo'].astype(str).str.startswith('C')]

In [22]:
after_cancel_drop = df.shape[0]
after_cancel_drop

397924

In [23]:
removed_cancel_rows = after_customer_id_drop - after_cancel_drop
removed_cancel_rows

8905

In [24]:
# Removed ~2% due to missing cancellations
(removed_cancel_rows / after_customer_id_drop) * 100

2.188880340388714

In [25]:
# 🔥 STEP 9 – Check Negative Quantity

In [26]:
# Checking Negative Quantites count
(df['Quantity'] <= 0).sum()

0

In [27]:
# 🔥 STEP 10 – Check UnitPrice ≤ 0

In [28]:
# Checking Negative Unitprice count
(df['UnitPrice'] <= 0).sum()

40

In [29]:
# 🔥 STEP 11 – Remove Zero or Negative UnitPrice

In [30]:
# Removeing Zero or Negative UnitPrice
df = df[df['UnitPrice'] > 0]

In [31]:
final_row_count = df.shape[0]
final_row_count

397884

In [32]:
# Removed 40 Unitprice rows
removed_price_rows = after_cancel_drop - final_row_count
removed_price_rows

40

In [33]:
# After removing:

# Missing CustomerID → 135,080 rows
# Cancellations → 8,905 rows
# Zero price rows → 40 rows

In [34]:
# converted CustomerID to integer.
df['CustomerID'] = df['CustomerID'].astype(int)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 397884 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    397884 non-null  object        
 1   StockCode    397884 non-null  object        
 2   Description  397884 non-null  object        
 3   Quantity     397884 non-null  int64         
 4   InvoiceDate  397884 non-null  datetime64[ns]
 5   UnitPrice    397884 non-null  float64       
 6   CustomerID   397884 non-null  int32         
 7   Country      397884 non-null  object        
dtypes: datetime64[ns](1), float64(1), int32(1), int64(1), object(4)
memory usage: 25.8+ MB


In [35]:
# 🔥 STEP 12 – Feature Engineering (Where Business Starts)

In [36]:
# Revenue = Quantity × UnitPrice
df['Revenue'] = df['Quantity'] * df['UnitPrice']
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Revenue
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,15.30
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34


In [37]:
df['Revenue'].describe()

count    397884.000000
mean         22.397000
std         309.071041
min           0.001000
25%           4.680000
50%          11.800000
75%          19.800000
max      168469.600000
Name: Revenue, dtype: float64

In [38]:
# 🔥 STEP 13 – Order-Level Aggregation

In [39]:
order_df = df.groupby(['InvoiceNo', 'CustomerID', 'InvoiceDate', 'Country'])['Revenue'].sum().reset_index()

order_df.head()

,InvoiceNo,CustomerID,InvoiceDate,Country,Revenue
0,536365,17850,2010-12-01 08:26:00,United Kingdom,139.12
1,536366,17850,2010-12-01 08:28:00,United Kingdom,22.20
2,536367,13047,2010-12-01 08:34:00,United Kingdom,278.73
3,536368,13047,2010-12-01 08:34:00,United Kingdom,70.05
4,536369,13047,2010-12-01 08:35:00,United Kingdom,17.85


In [40]:
order_df.shape

(18562, 5)

In [41]:
# 🔥 STEP 14 – Basic Order-Level Insights

In [42]:
order_df['Revenue'].describe()

count     18562.000000
mean        480.088778
std        1676.772723
min           0.380000
25%         157.700000
50%         302.875000
75%         471.257500
max      168469.600000
Name: Revenue, dtype: float64

In [43]:
order_df['CustomerID'].nunique()

4338

In [44]:
# 🔥 STEP 15 – Customer-Level Aggregation

In [45]:
customer_df = order_df.groupby('CustomerID').agg(
    Total_Revenue=('Revenue', 'sum'),
    Total_Orders=('InvoiceNo', 'count'),
    First_Purchase=('InvoiceDate', 'min'),
    Last_Purchase=('InvoiceDate', 'max')
).reset_index()

customer_df.head()

,CustomerID,Total_Revenue,Total_Orders,First_Purchase,Last_Purchase
0,12346,77183.60,1,2011-01-18 10:01:00,2011-01-18 10:01:00
1,12347,4310.00,7,2010-12-07 14:57:00,2011-12-07 15:52:00
2,12348,1797.24,4,2010-12-16 19:09:00,2011-09-25 13:13:00
3,12349,1757.55,1,2011-11-21 09:51:00,2011-11-21 09:51:00
4,12350,334.40,1,2011-02-02 16:01:00,2011-02-02 16:01:00


In [46]:
customer_df.shape

(4338, 5)

In [47]:
# 🔥 STEP 16 – Create Recency Metric

In [48]:
max_date = order_df['InvoiceDate'].max()
max_date

Timestamp('2011-12-09 12:50:00')

In [49]:
customer_df['Recency_Days'] = (max_date - customer_df['Last_Purchase']).dt.days
customer_df.head()

,CustomerID,Total_Revenue,Total_Orders,First_Purchase,Last_Purchase,Recency_Days
0,12346,77183.60,1,2011-01-18 10:01:00,2011-01-18 10:01:00,325
1,12347,4310.00,7,2010-12-07 14:57:00,2011-12-07 15:52:00,1
2,12348,1797.24,4,2010-12-16 19:09:00,2011-09-25 13:13:00,74
3,12349,1757.55,1,2011-11-21 09:51:00,2011-11-21 09:51:00,18
4,12350,334.40,1,2011-02-02 16:01:00,2011-02-02 16:01:00,309


In [50]:
# 🔥 STEP 17 – Confirm Frequency & Monetary

In [51]:
customer_df[['Total_Orders', 'Total_Revenue', 'Recency_Days']].describe()

,Total_Orders,Total_Revenue,Recency_Days
count,4338.000000,4338.000000,4338.000000
mean,4.278930,2054.266460,91.536422
std,7.720123,8989.230441,100.014169
min,1.000000,3.750000,0.000000
25%,1.000000,307.415000,17.000000
50%,2.000000,674.485000,50.000000
75%,5.000000,1661.740000,141.000000
max,210.000000,280206.020000,373.000000


In [52]:
# # 🔥 STEP 18 – Create RFM Scores
# We will create:
# R_Score (lower Recency = better)
# F_Score (higher Frequency = better)
# M_Score (higher Monetary = better)

In [53]:
# Recency Score (reverse order)
customer_df['R_Score'] = pd.qcut(customer_df['Recency_Days'], 5, labels=[5,4,3,2,1])

# Frequency Score
customer_df['F_Score'] = pd.qcut(customer_df['Total_Orders'].rank(method='first'), 5, labels=[1,2,3,4,5])

# Monetary Score
customer_df['M_Score'] = pd.qcut(customer_df['Total_Revenue'], 5, labels=[1,2,3,4,5])

In [54]:
customer_df[['CustomerID','R_Score','F_Score','M_Score']].head()

,CustomerID,R_Score,F_Score,M_Score
0,12346,1,1,5
1,12347,5,5,5
2,12348,2,4,4
3,12349,4,1,4
4,12350,1,1,2


In [55]:
# 🔥 STEP 19 – Create Combined RFM Score

In [56]:
customer_df['RFM_Score'] = (
    customer_df['R_Score'].astype(str) +
    customer_df['F_Score'].astype(str) +
    customer_df['M_Score'].astype(str)
)

customer_df[['CustomerID','RFM_Score']].head()

,CustomerID,RFM_Score
0,12346,115
1,12347,555
2,12348,244
3,12349,414
4,12350,112


In [57]:
customer_df['RFM_Score'].value_counts().head(10)

RFM_Score
555    348
111    183
455    178
121    153
112    128
444    122
122     91
544     89
233     86
344     84
Name: count, dtype: int64

In [58]:
# 🔥 STEP 20 – Create RFM Segments

In [59]:
def segment_customer(row):
    if row['RFM_Score'] == '555':
        return 'Champions'
    elif row['R_Score'] >= 4 and row['F_Score'] >= 4:
        return 'Loyal Customers'
    elif row['R_Score'] <= 2 and row['F_Score'] >= 4:
        return 'At Risk'
    elif row['R_Score'] >= 4 and row['F_Score'] <= 2:
        return 'Potential Loyalist'
    elif row['R_Score'] <= 2 and row['F_Score'] <= 2:
        return 'Lost'
    else:
        return 'Others'

customer_df['Segment'] = customer_df.apply(segment_customer, axis=1)

customer_df['Segment'].value_counts()

Segment
Others                1542
Lost                  1064
Loyal Customers        790
Champions              348
Potential Loyalist     320
At Risk                274
Name: count, dtype: int64

**🎯 Now We Validate Pareto Principle**

We test:
Do top customers generate most revenue?

In [60]:
# Sort customers by revenue
top_customers = customer_df.sort_values(by='Total_Revenue', ascending=False)

# Calculate cumulative revenue percentage
top_customers['Revenue_%'] = top_customers['Total_Revenue'] / top_customers['Total_Revenue'].sum()
top_customers['Cumulative_%'] = top_customers['Revenue_%'].cumsum()

top_customers.head()

,CustomerID,Total_Revenue,Total_Orders,First_Purchase,Last_Purchase,Recency_Days,R_Score,F_Score,M_Score,RFM_Score,Segment,Revenue_%,Cumulative_%
1689,14646,280206.02,73,2010-12-20 10:09:00,2011-12-08 12:12:00,1,5,5,5,555,Champions,0.031444,0.031444
4201,18102,259657.30,60,2010-12-07 16:42:00,2011-12-09 11:50:00,0,5,5,5,555,Champions,0.029138,0.060581
3728,17450,194550.79,46,2010-12-07 09:23:00,2011-12-01 13:29:00,7,5,5,5,555,Champions,0.021832,0.082413
3008,16446,168472.50,2,2011-05-18 09:52:00,2011-12-09 09:15:00,0,5,3,5,535,Others,0.018905,0.101318
1879,14911,143825.06,202,2010-12-01 14:05:00,2011-12-08 15:54:00,0,5,5,5,555,Champions,0.016139,0.117457


In [61]:
top_customers[top_customers['Cumulative_%'] <= 0.8].shape

(1132, 13)

**🔥 Now We Validate Something More Powerful**

We need to check:

👉 Do Champions overlap with top revenue customers?

In [62]:
top_80 = top_customers[top_customers['Cumulative_%'] <= 0.8]

top_80['Segment'].value_counts()

Segment
Loyal Customers       392
Champions             348
Others                251
At Risk                97
Lost                   34
Potential Loyalist     10
Name: count, dtype: int64

**🔥 Next Move (Important)**

Now we quantify revenue by segment.

In [63]:
segment_revenue = customer_df.groupby('Segment')['Total_Revenue'].sum().sort_values(ascending=False)

segment_revenue

Segment
Champions             3905165.910
Loyal Customers       2022212.541
Others                1883287.831
Lost                   518984.512
At Risk                435150.620
Potential Loyalist     146606.490
Name: Total_Revenue, dtype: float64

In [64]:
segment_revenue / segment_revenue.sum()

Segment
Champions             0.438221
Loyal Customers       0.226924
Others                0.211334
Lost                  0.058238
At Risk               0.048831
Potential Loyalist    0.016452
Name: Total_Revenue, dtype: float64

📊 Revenue Contribution by Segment

Champions → 43.8%

Loyal Customers → 22.7%

Others → 21.1%

Lost → 5.8%

At Risk → 4.9%

Potential Loyalist → 1.6%

**🚀 Now We Move To Advanced KPI**

**Next we calculate:**  Repeat Purchase Rate

In [65]:
repeat_rate = (customer_df['Total_Orders'] > 1).mean()

repeat_rate

0.6562932226832642

**🧠 Strategic Interpretation**

This means:  

Customer acquisition cost is recoverable
Brand loyalty exists
Email/remarketing likely effective
CLV likely strong

But here’s the real question:

👉 Is repeat rate driven only by Champions? Or across segments?

Now we test segment-level repeat rate.

In [66]:
customer_df.groupby('Segment')['Total_Orders'].apply(lambda x: (x > 1).mean())

Segment
At Risk               1.000000
Champions             1.000000
Lost                  0.102444
Loyal Customers       1.000000
Others                0.809339
Potential Loyalist    0.243750
Name: Total_Orders, dtype: float64

**Where is growth potential?**

Average Revenue per Customer by Segment.

In [67]:
customer_df.groupby('Segment')['Total_Revenue'].mean().sort_values(ascending=False)

Segment
Champions             11221.741121
Loyal Customers        2559.762710
At Risk                1588.140949
Others                 1221.328036
Lost                    487.767398
Potential Loyalist      458.145281
Name: Total_Revenue, dtype: float64

**✅ Correct Strategic Choice:**

At Risk segment

Because:

High past value, High repeat behavior, Recently cooling, High recovery probability ,This is textbook reactivation ROI logic.

**🧠 This Is How Analysts Think**

Not: “Which group is biggest?”

But: “Where is marginal revenue per marketing dollar highest?”

Now next step.

We quantify revenue at risk.

In [68]:
customer_df[customer_df['Segment'] == 'At Risk']['Total_Revenue'].sum()

435150.62

**🚀 Now We Move To Advanced KPI**

We now calculate: Customer Lifetime Value(CLV) (Simple Model)

In [69]:
customer_df['CLV'] = customer_df['Total_Revenue']

customer_df['CLV'].describe()

count      4338.000000
mean       2054.266460
std        8989.230441
min           3.750000
25%         307.415000
50%         674.485000
75%        1661.740000
max      280206.020000
Name: CLV, dtype: float64

Average Order Value (AOV)

In [70]:
aov = order_df['Revenue'].mean()

aov

480.08877836440035

What % of revenue comes from top 10% customers?

In [71]:
top_10_percent = int(0.1 * len(customer_df))

top_10_revenue = customer_df.sort_values(
    by='Total_Revenue', ascending=False
).head(top_10_percent)['Total_Revenue'].sum()

top_10_revenue / customer_df['Total_Revenue'].sum()

0.6133411711012168

🚀 Now Final Strategic Question

If top 10% generate 61% revenue,
What % does bottom 50% generate?

In [72]:
bottom_50 = customer_df.sort_values(
    by='Total_Revenue', ascending=True
).head(int(0.5 * len(customer_df)))['Total_Revenue'].sum()

bottom_50 / customer_df['Total_Revenue'].sum()

0.07832285869067991

🎯 STEP 1 — Decide Warehouse Design

In [73]:
order_df.head()

,InvoiceNo,CustomerID,InvoiceDate,Country,Revenue
0,536365,17850,2010-12-01 08:26:00,United Kingdom,139.12
1,536366,17850,2010-12-01 08:28:00,United Kingdom,22.20
2,536367,13047,2010-12-01 08:34:00,United Kingdom,278.73
3,536368,13047,2010-12-01 08:34:00,United Kingdom,70.05
4,536369,13047,2010-12-01 08:35:00,United Kingdom,17.85


🚀 STEP 2 — Create Date Dimension in Python

In [74]:
date_df = pd.DataFrame({
    'FullDate': pd.to_datetime(order_df['InvoiceDate'].dt.date.unique())
})

date_df['DateKey'] = date_df['FullDate'].dt.strftime('%Y%m%d').astype(int)
date_df['Year'] = date_df['FullDate'].dt.year
date_df['Month'] = date_df['FullDate'].dt.month
date_df['MonthName'] = date_df['FullDate'].dt.strftime('%B')
date_df['Quarter'] = date_df['FullDate'].dt.quarter

date_df.head()

,FullDate,DateKey,Year,Month,MonthName,Quarter
0,2010-12-01,20101201,2010,12,December,4
1,2010-12-02,20101202,2010,12,December,4
2,2010-12-03,20101203,2010,12,December,4
3,2010-12-05,20101205,2010,12,December,4
4,2010-12-06,20101206,2010,12,December,4


🔥 STEP 3 — Prepare Fact Table Properly

In [75]:
order_df['DateKey'] = order_df['InvoiceDate'].dt.strftime('%Y%m%d').astype(int)

fact_invoices = order_df[[
    'InvoiceNo',
    'CustomerID',
    'DateKey',
    'Country',
    'Revenue'
]].copy()

fact_invoices.head()

,InvoiceNo,CustomerID,DateKey,Country,Revenue
0,536365,17850,20101201,United Kingdom,139.12
1,536366,17850,20101201,United Kingdom,22.20
2,536367,13047,20101201,United Kingdom,278.73
3,536368,13047,20101201,United Kingdom,70.05
4,536369,13047,20101201,United Kingdom,17.85


🔥 STEP 4 — Build Customer Dimension

Right now customer_df has extra analytical columns.

We need to turn it into dim_customers.

In [76]:
dim_customers = customer_df[[
    'CustomerID',
    'Total_Revenue',
    'Total_Orders',
    'Recency_Days',
    'RFM_Score',
    'Segment'
]].copy()

dim_customers.head()

,CustomerID,Total_Revenue,Total_Orders,Recency_Days,RFM_Score,Segment
0,12346,77183.60,1,325,115,Lost
1,12347,4310.00,7,1,555,Champions
2,12348,1797.24,4,74,244,At Risk
3,12349,1757.55,1,18,414,Potential Loyalist
4,12350,334.40,1,309,112,Lost


🔥 STEP 5 — Validate Model Integrity (Very Important)

Before exporting, we test something most beginners skip.

We must confirm:

1️⃣ No duplicate customers in dim_customers

In [77]:
dim_customers['CustomerID'].nunique(), len(dim_customers)

(4338, 4338)

🔥 STEP 6 — Validate Fact Table Link

Now we confirm:

Every CustomerID in fact_invoices exists in dim_customers.

In [78]:
fact_invoices['CustomerID'].isin(dim_customers['CustomerID']).all()

True

🔥 STEP 7 — Export Tables (Move to SQL World)

Now we export the three tables.

In [80]:
fact_invoices.to_csv("fact_invoices.csv", index=False)
dim_customers.to_csv("dim_customers.csv", index=False)
date_df.to_csv("dim_date.csv", index=False)